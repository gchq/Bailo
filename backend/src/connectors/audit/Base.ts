import { Request } from 'express'

import { AccessRequestDoc } from '../../models/AccessRequest.js'
import { FileInterface, FileInterfaceDoc } from '../../models/File.js'
import { InferenceDoc } from '../../models/Inference.js'
import { ModelCardInterface, ModelDoc, ModelInterface } from '../../models/Model.js'
import { ReleaseDoc } from '../../models/Release.js'
import { ResponseInterface } from '../../models/Response.js'
import { ReviewInterface } from '../../models/Review.js'
import { SchemaDoc, SchemaInterface } from '../../models/Schema.js'
import { TokenDoc } from '../../models/Token.js'
import { ModelSearchResult } from '../../routes/v2/model/getModelsSearch.js'
import { BailoError } from '../../types/error.js'

const AuditKind = {
  Create: 'Create',
  View: 'View',
  Update: 'Update',
  Delete: 'Delete',
  Search: 'Search',
} as const
export type AuditKindKeys = (typeof AuditKind)[keyof typeof AuditKind]

export const AuditInfo = {
  CreateModel: { typeId: 'CreateModel', description: 'Model Created', auditKind: AuditKind.Create },
  ViewModel: { typeId: 'ViewModel', description: 'Model Viewed', auditKind: AuditKind.View },
  UpdateModel: { typeId: 'UpdateModel', description: 'Model Updated', auditKind: AuditKind.Update },
  SearchModels: { typeId: 'SearchModels', description: 'Model Searched', auditKind: AuditKind.Search },

  CreateModelCard: { typeId: 'CreateModelCard', description: 'Model Card Created', auditKind: AuditKind.Create },
  ViewModelCard: { typeId: 'ViewModelCard', description: 'Model Card Viewed', auditKind: AuditKind.View },
  ViewModelCardRevisions: {
    typeId: 'ViewModelCardRevisions',
    description: 'Model Card Revisions Viewed',
    auditKind: AuditKind.View,
  },
  UpdateModelCard: { typeId: 'UpdateModelCard', description: 'Model Card Updated', auditKind: AuditKind.Update },

  CreateFile: { typeId: 'CreateFile', description: 'File Information Created', auditKind: AuditKind.Create },
  ViewFile: { typeId: 'ViewFile', description: 'File Downloaded', auditKind: AuditKind.View },
  ViewFiles: { typeId: 'ViewFiles', description: 'File Information Viewed', auditKind: AuditKind.View },
  DeleteFile: { typeId: 'DeleteFile', description: 'File Information Deleted', auditKind: AuditKind.Delete },

  CreateRelease: { typeId: 'CreateRelease', description: 'Release Created', auditKind: AuditKind.Create },
  ViewRelease: { typeId: 'ViewRelease', description: 'Release Viewed', auditKind: AuditKind.View },
  UpdateRelease: { typeId: 'UpdateRelease', description: 'Release Updated', auditKind: AuditKind.Update },
  DeleteRelease: { typeId: 'DeleteRelease', description: 'Release Deleted', auditKind: AuditKind.Delete },
  ViewReleases: { typeId: 'ViewReleases', description: 'Releases Viewed', auditKind: AuditKind.View },

  CreateUserToken: { typeId: 'CreateUserToken', description: 'Token Created', auditKind: AuditKind.Create },
  ViewUserTokens: { typeId: 'ViewUserToken', description: 'Token Viewed', auditKind: AuditKind.View },
  DeleteUserToken: { typeId: 'DeleteUserToken', description: 'Token Deleted', auditKind: AuditKind.Delete },

  CreateAccessRequest: {
    typeId: 'CreateAccessRequest',
    description: 'Access Request Created',
    auditKind: AuditKind.Create,
  },
  ViewAccessRequest: { typeId: 'ViewAccessRequest', description: 'Access Request Viewed', auditKind: AuditKind.View },
  UpdateAccessRequest: {
    typeId: 'UpdateAccess Request',
    description: 'Access Request Updated',
    auditKind: AuditKind.Update,
  },
  DeleteAccessRequest: {
    typeId: 'UpdateAccessRequest',
    description: 'Access Request Deleted',
    auditKind: AuditKind.Update,
  },
  ViewAccessRequests: {
    typeId: 'ViewAccessRequests',
    description: 'Access Requests Viewed',
    auditKind: AuditKind.View,
  },

  SearchReviews: {
    typeId: 'SearchReviews',
    description: 'Reviews Searched',
    auditKind: AuditKind.Search,
  },
  CreateReviewResponse: {
    typeId: 'CreateReviewResponse',
    description: 'Review Response Created',
    auditKind: AuditKind.Create,
  },
  UpdateReviewResponse: {
    typeId: 'UpdateReviewResponse',
    description: 'Review Response Updated',
    auditKind: AuditKind.Update,
  },

  CreateSchema: { typeId: 'CreateSchema', description: 'Schema Created', auditKind: AuditKind.Create },
  SearchSchemas: { typeId: 'SearchedSchemas', description: 'Schemas Searched', auditKind: AuditKind.Search },
  ViewSchema: { typeId: 'ViewSchema', description: 'Schema Viewed', auditKind: AuditKind.View },
  DeleteSchema: { typeId: 'DeleteSchema', description: 'Schema Deleted', auditKind: AuditKind.Delete },
  UpdateSchema: { typeId: 'UpdateSchema', description: 'Schema Updated', auditKind: AuditKind.Update },

  ViewModelImages: { typeId: 'ViewModelImages', description: 'Model Images Viewed', auditKind: AuditKind.View },

  CreateInference: { typeId: 'CreateInference', description: 'Inference Service Created', auditKind: AuditKind.Create },
  UpdateInference: { typeId: 'UpdateInference', description: 'Inference Service Updated', auditKind: AuditKind.Update },
  ViewInference: { typeId: 'ViewInference', description: 'Inference Service Viewed', auditKind: AuditKind.View },
  ViewInferences: { typeId: 'ViewInferences', description: 'Inferences Viewed', auditKind: AuditKind.View },

  CreateExport: { typeId: 'CreateExport', description: 'Model Exported', auditKind: AuditKind.Create },
  CreateImport: { typeId: 'CreateImport', description: 'Model Imported', auditKind: AuditKind.Create },

  ViewResponses: {
    typeId: 'ViewResponses',
    description: 'View a list of responses',
    auditKind: AuditKind.View,
  },
  CreateResponse: {
    typeId: 'CreateResponse',
    description: 'Review or comment responses created',
    auditKind: AuditKind.Create,
  },
  UpdateResponse: {
    typeId: 'UpdateResponse',
    description: 'Updated a comment or review response',
    auditKind: AuditKind.Update,
  },
} as const
export type AuditInfoKeys = (typeof AuditInfo)[keyof typeof AuditInfo]

export abstract class BaseAuditConnector {
  abstract onCreateModel(req: Request, model: ModelDoc)
  abstract onViewModel(req: Request, model: ModelDoc)
  abstract onUpdateModel(req: Request, model: ModelDoc)
  abstract onSearchModel(req: Request, models: ModelSearchResult[])

  abstract onCreateModelCard(req: Request, modelId: string, modelCard: ModelCardInterface)
  abstract onViewModelCard(req: Request, modelId: string, modelCard: ModelCardInterface)
  abstract onUpdateModelCard(req: Request, modelId: string, modelCard: ModelCardInterface)
  abstract onViewModelCardRevisions(req: Request, modelId: string, modelCards: ModelCardInterface[])

  abstract onCreateFile(req: Request, file: FileInterfaceDoc)
  abstract onViewFile(req: Request, file: FileInterfaceDoc)
  abstract onViewFiles(req: Request, modelId: string, files: FileInterface[])
  abstract onDeleteFile(req: Request, modelId: string, fileId: string)

  abstract onCreateRelease(req: Request, release: ReleaseDoc)
  abstract onViewRelease(req: Request, release: ReleaseDoc)
  abstract onUpdateRelease(req: Request, release: ReleaseDoc)
  abstract onDeleteRelease(req: Request, modelId: string, semver: string)
  abstract onViewReleases(req: Request, releases: ReleaseDoc[])

  abstract onCreateCommentResponse(req: Request, response: ResponseInterface)
  abstract onCreateReviewResponse(req: Request, response: ResponseInterface)

  abstract onViewResponses(req: Request, responseInterfaces: ResponseInterface[])
  abstract onUpdateResponse(req: Request, responseId: string)

  abstract onCreateUserToken(req: Request, token: TokenDoc)
  abstract onViewUserTokens(req: Request, tokens: TokenDoc[])
  abstract onDeleteUserToken(req: Request, accessKey: string)

  abstract onCreateAccessRequest(req: Request, accessRequest: AccessRequestDoc)
  abstract onViewAccessRequest(req: Request, accessRequest: AccessRequestDoc)
  abstract onUpdateAccessRequest(req: Request, accessRequest: AccessRequestDoc)
  abstract onDeleteAccessRequest(req: Request, accessRequestId: string)
  abstract onViewAccessRequests(req: Request, accessRequests: AccessRequestDoc[])

  abstract onSearchReviews(req: Request, reviews: (ReviewInterface & { model: ModelInterface })[])

  abstract onSearchSchemas(req: Request, schemas: SchemaInterface[])
  abstract onCreateSchema(req: Request, schema: SchemaInterface)
  abstract onViewSchema(req: Request, schema: SchemaInterface)
  abstract onDeleteSchema(req: Request, schemaId: string)
  abstract onUpdateSchema(req: Request, schema: SchemaDoc)

  abstract onCreateInference(req: Request, inference: InferenceDoc)
  abstract onUpdateInference(req: Request, inference: InferenceDoc)
  abstract onViewInference(req: Request, inference: InferenceDoc)
  abstract onViewInferences(req: Request, inference: InferenceDoc[])

  abstract onViewModelImages(
    req: Request,
    modelId: string,
    images: { repository: string; name: string; tags: string[] }[],
  )

  abstract onCreateS3Export(req: Request, modelId: string, semvers?: string[])
  abstract onCreateImport(
    req: Request,
    mirroredModelId: string,
    sourceModelId: string,
    modelCardVersions: number[],
    exporter: string,
  )

  abstract onError(req: Request, error: BailoError)

  checkEventType(auditInfo: AuditInfoKeys, req: Request) {
    if (auditInfo.typeId !== req.audit.typeId && auditInfo.description !== req.audit.description) {
      throw new Error(`Audit: Expected type '${JSON.stringify(auditInfo)}' but recieved '${JSON.stringify(req.audit)}'`)
    }
  }
}
