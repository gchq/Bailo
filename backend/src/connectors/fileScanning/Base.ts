import { FileInterface } from '../../models/File.js'
import { ScanInterface } from '../../models/Scan.js'
import log from '../../services/log.js'
export type FileScanResult = Pick<
  ScanInterface,
  'toolName' | 'scannerVersion' | 'state' | 'isInfected' | 'viruses' | 'lastRunAt'
>

export const ScanState = {
  NotScanned: 'notScanned',
  InProgress: 'inProgress',
  Complete: 'complete',
  Error: 'error',
} as const
export type ScanStateKeys = (typeof ScanState)[keyof typeof ScanState]

export type FileScanningConnectorInfo = Pick<FileScanResult, 'toolName' | 'scannerVersion'>

export abstract class BaseFileScanningConnector {
  abstract toolName: string
  abstract version: string | undefined

  abstract init()
  abstract scan(file: FileInterface): Promise<FileScanResult[]>

  info(): FileScanningConnectorInfo {
    return { toolName: this.toolName, scannerVersion: this.version }
  }

  async scanError(message: string, context?: object): Promise<FileScanResult[]> {
    const scannerInfo = await this.info()
    log.error({ ...context, ...scannerInfo }, message)
    return [
      {
        ...scannerInfo,
        state: ScanState.Error,
        lastRunAt: new Date(),
      },
    ]
  }
}
