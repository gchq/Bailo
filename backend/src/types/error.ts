import Logger from 'bunyan'

export interface BailoError extends Error {
  // Inherited from 'Error'
  // name: string
  // message: string

  // An HTTP response code that represents the error
  name: string
  message: string
  code: number
  status: number

  // This data is logged publicly to the frontend
  context?: {
    documentationUrl?: string

    // Items in internal are not displayed to the user and may include sensitive information
    internal?: unknown

    [key: string]: unknown
  }

  // A custom logger may be provided, otherwise a default is used
  logger?: Logger
}

export function isBailoError(err: unknown): err is BailoError {
  if (typeof err !== 'object' || err === null) {
    return false
  }

  if (err instanceof Error && err.name === 'Bailo Error') {
    return true
  }

  return false
}
