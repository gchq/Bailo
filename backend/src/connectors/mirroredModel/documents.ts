import { PassThrough } from 'node:stream'
import { json } from 'node:stream/consumers'

import { ObjectId } from 'mongoose'
import { Headers } from 'tar-stream'

import { ReleaseAction } from '../../connectors/authorisation/actions.js'
import authorisation from '../../connectors/authorisation/index.js'
import { ModelDoc } from '../../models/Model.js'
import { ModelCardRevisionDoc } from '../../models/ModelCardRevision.js'
import { ReleaseDoc } from '../../models/Release.js'
import { UserInterface } from '../../models/User.js'
import { saveImportedFile } from '../../services/file.js'
import log from '../../services/log.js'
import { parseFile, parseModelCard, parseRelease } from '../../services/mirroredModel/entityParsers.js'
import { getModelById, saveImportedModelCard, setLatestImportedModelCard } from '../../services/model.js'
import { DistributionPackageName, joinDistributionPackageName } from '../../services/registry.js'
import { saveImportedRelease } from '../../services/release.js'
import config from '../../utils/config.js'
import { Forbidden, InternalError } from '../../utils/error.js'
import { BaseImporter, BaseMirrorMetadata } from './base.js'
import { MirrorKind, MirrorKindKeys } from './index.js'

export type DocumentsMirrorMetadata = BaseMirrorMetadata & { importKind: MirrorKindKeys<'Documents'> }
export type MongoDocumentMirrorInformation = {
  metadata: DocumentsMirrorMetadata
  modelCardVersions: ModelCardRevisionDoc['version'][]
  newModelCards: Omit<ModelCardRevisionDoc, '_id'>[]
  releaseSemvers: ReleaseDoc['semver'][]
  newReleases: Omit<ReleaseDoc, '_id'>[]
  fileIds: ObjectId[]
  imageIds: string[]
}

export class DocumentsImporter extends BaseImporter {
  declare protected metadata: DocumentsMirrorMetadata

  protected user: UserInterface

  protected modelCardVersions: number[] = []
  protected releaseSemvers: string[] = []
  protected fileIds: ObjectId[] = []
  protected imageIds: string[] = []
  protected newModelCards: Omit<ModelCardRevisionDoc, '_id'>[] = []
  protected newReleases: Omit<ReleaseDoc, '_id'>[] = []

  protected lazyMirroredModel: ModelDoc | null = null
  protected modelCardRegex = new RegExp(String.raw`^${config.modelMirror.contentDirectory}/[0-9]+\.json$`)
  protected releaseRegex = new RegExp(String.raw`^${config.modelMirror.contentDirectory}/releases\/(.*)\.json$`)
  protected fileRegex = new RegExp(String.raw`^${config.modelMirror.contentDirectory}/files\/(.*)\.json$`)

  constructor(user: UserInterface, metadata: DocumentsMirrorMetadata, logData?: Record<string, unknown>) {
    super(metadata, logData)
    if (this.metadata.importKind !== MirrorKind.Documents) {
      throw InternalError('Cannot parse compressed Documents: incorrect metadata specified.', {
        metadata: this.metadata,
        ...this.logData,
      })
    }
    this.user = user
  }

  async processEntry(entry: Headers, stream: PassThrough) {
    if (entry.type === 'file') {
      // Process file
      const fileContentsJson = await json(stream)
      if (this.modelCardRegex.test(entry.name)) {
        log.debug(
          { name: entry.name, metadata: this.metadata, ...this.logData },
          'Processing compressed file as a Model Card.',
        )
        const modelCard = parseModelCard(fileContentsJson, this.metadata.mirroredModelId, this.metadata.sourceModelId, {
          metadata: this.metadata,
          ...this.logData,
        })
        this.modelCardVersions.push(modelCard.version)
        const savedModelCard = await saveImportedModelCard(modelCard)
        if (savedModelCard) {
          this.newModelCards.push(savedModelCard)
        }
      } else if (this.releaseRegex.test(entry.name)) {
        log.debug(
          { name: entry.name, metadata: this.metadata, ...this.logData },
          'Processing compressed file as a Release.',
        )
        const release = parseRelease(fileContentsJson, this.metadata.mirroredModelId, this.metadata.sourceModelId, {
          metadata: this.metadata,
          ...this.logData,
        })

        // have to authorise per-release due to streaming, rather than do all releases at once
        if (!this.lazyMirroredModel) {
          this.lazyMirroredModel = await getModelById(this.user, this.metadata.mirroredModelId)
        }
        const authResponse = await authorisation.releases(
          this.user,
          this.lazyMirroredModel,
          [release],
          ReleaseAction.Import,
        )
        if (!authResponse[0]?.success) {
          throw Forbidden('Insufficient permissions to import the specified releases.', {
            modelId: this.metadata.mirroredModelId,
            release: release.semver,
            user: this.user,
            metadata: this.metadata,
            ...this.logData,
          })
        }

        for (const image of release.images) {
          this.imageIds.push(
            joinDistributionPackageName({
              domain: image.repository,
              path: image.name,
              tag: image.tag,
            } as DistributionPackageName),
          )
        }
        this.releaseSemvers.push(release.semver)
        const savedRelease = await saveImportedRelease(release)
        if (savedRelease) {
          this.newReleases.push(savedRelease)
        }
      } else if (this.fileRegex.test(entry.name)) {
        log.debug(
          { name: entry.name, metadata: this.metadata, ...this.logData },
          'Processing compressed file as a File.',
        )
        const file = await parseFile(fileContentsJson, this.metadata.mirroredModelId, this.metadata.sourceModelId, {
          metadata: this.metadata,
          ...this.logData,
        })
        this.fileIds.push(file._id)
        await saveImportedFile(file)
      } else {
        throw InternalError('Cannot parse compressed file: unrecognised contents.', {
          metadata: this.metadata,
          ...this.logData,
        })
      }
    } else {
      // skip entry of type: link | symlink | directory | block-device | character-device | fifo | contiguous-file
      log.debug(
        { name: entry.name, type: entry.type, metadata: this.metadata, ...this.logData },
        'Skipping non-file entry.',
      )
    }
  }

  async finishListener(
    resolve: (reason?: MongoDocumentMirrorInformation) => void,
    _reject: (reason?: unknown) => void,
    _logData?: Record<string, unknown>,
  ) {
    log.debug({ metadata: this.metadata, ...this.logData }, 'Completed extracting archive.')

    await setLatestImportedModelCard(this.metadata.mirroredModelId)

    log.info(
      {
        mirroredModelId: this.metadata.mirroredModelId,
        sourceModelId: this.metadata.sourceModelId,
        modelCardVersions: this.modelCardVersions,
        releaseSemvers: this.releaseSemvers,
        fileIds: this.fileIds,
        imageIds: this.imageIds,
        numberOfDocuments: {
          modelCards: this.newModelCards.length,
          releases: this.newReleases.length,
          files: this.fileIds.length,
          images: this.imageIds.length,
        },
        metadata: this.metadata,
        ...this.logData,
      },
      'Finished importing the collection of model documents.',
    )

    resolve({
      metadata: this.metadata,
      modelCardVersions: this.modelCardVersions,
      newModelCards: this.newModelCards,
      releaseSemvers: this.releaseSemvers,
      newReleases: this.newReleases,
      fileIds: this.fileIds,
      imageIds: this.imageIds,
    })
  }
}
