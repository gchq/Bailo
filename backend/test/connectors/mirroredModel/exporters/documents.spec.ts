import { PassThrough } from 'node:stream'

import { beforeEach, describe, expect, test, vi } from 'vitest'

import { ScanState } from '../../../../src/connectors/fileScanning/Base.js'
import { DocumentsExporter } from '../../../../src/connectors/mirroredModel/exporters/documents.js'
import { MirrorKind } from '../../../../src/connectors/mirroredModel/index.js'
import { BadReq, InternalError } from '../../../../src/utils/error.js'

const tarballMocks = vi.hoisted(() => ({
  initialiseTarGzUpload: vi.fn(),
  addEntryToTarGzUpload: vi.fn(),
}))
vi.mock('../../../../src/services/mirroredModel/tarball.js', () => tarballMocks)

const fileServiceMocks = vi.hoisted(() => ({
  getFilesByIds: vi.fn(),
  getTotalFileSize: vi.fn(),
}))
vi.mock('../../../../src/services/file.js', () => fileServiceMocks)

const modelServiceMocks = vi.hoisted(() => ({
  getModelCardRevisions: vi.fn(),
}))
vi.mock('../../../../src/services/model.js', () => modelServiceMocks)

const releaseServiceMocks = vi.hoisted(() => ({
  getAllFileIds: vi.fn(),
}))
vi.mock('../../../../src/services/release.js', () => releaseServiceMocks)

const logMocks = vi.hoisted(() => ({
  default: {
    debug: vi.fn(),
  },
}))
vi.mock('../../../../src/services/log.js', () => logMocks)

const scannersMocks = vi.hoisted(() => ({
  default: {
    info: vi.fn(),
  },
}))
vi.mock('../../../../src/connectors/fileScanning/index.js', () => scannersMocks)

const configMocks = vi.hoisted(() => ({
  default: {
    modelMirror: {
      export: {
        maxSize: 1000,
      },
    },
  },
}))
vi.mock('../../../../src/utils/config.js', () => configMocks)

const authMocks = vi.hoisted(() => ({
  default: {
    model: vi.fn(),
  },
}))
vi.mock('../../../../src/connectors/authorisation/index.js', () => authMocks)

const mockUser = { dn: 'userDN' } as any
const mockModel = {
  id: 'modelId',
  settings: { mirror: { destinationModelId: 'destModelId' } },
  card: { schemaId: 'schemaId' },
} as any

const mockRelease = {
  id: 'relId',
  semver: '1.0.0',
  modelId: 'modelId',
  fileIds: ['f1'],
  toJSON: () => ({ foo: 'bar' }),
} as any

const mockFile = {
  id: 'fileId',
  _id: { toString: () => 'fileId' },
  name: 'f',
  avScan: [{ state: ScanState.Complete, isInfected: false }],
} as any

describe('connectors > mirroredModel > exporters > DocumentsExporter', () => {
  beforeEach(() => {
    tarballMocks.initialiseTarGzUpload.mockResolvedValue({
      tarStream: {} as any,
      gzipStream: {} as any,
      uploadStream: {} as any,
      uploadPromise: Promise.resolve(),
    })
    tarballMocks.addEntryToTarGzUpload.mockResolvedValue(undefined)
    fileServiceMocks.getFilesByIds.mockResolvedValue([mockFile])
    fileServiceMocks.getTotalFileSize.mockResolvedValue(500)
    modelServiceMocks.getModelCardRevisions.mockResolvedValue([{ version: 'v1', toJSON: () => ({}) }])
    releaseServiceMocks.getAllFileIds.mockResolvedValue(['fileId'])
    scannersMocks.default.info.mockReturnValue(false)
    authMocks.default.model.mockResolvedValue({ success: true })
  })

  test('constructor > sets releases and logData', () => {
    const exporter = new DocumentsExporter(mockUser, mockModel, [mockRelease])

    expect(exporter.getReleases()).toEqual([mockRelease])
    expect(exporter).toMatchSnapshot()
  })

  test('_init > success without releases', async () => {
    const exporter = new DocumentsExporter(mockUser, mockModel)

    // @ts-expect-error protected method
    await exporter._init()

    expect(authMocks.default.model).toHaveBeenCalled()
  })

  test('_init > success with releases calls checkReleaseFiles', async () => {
    const exporter = new DocumentsExporter(mockUser, mockModel, [mockRelease])

    // @ts-expect-error protected method
    await exporter._init()

    expect(exporter.getFiles()).toEqual([mockFile])
  })

  test('checkReleaseFiles > returns early when multiple releases have no fileIds', async () => {
    const exporter = new DocumentsExporter(mockUser, mockModel, [mockRelease, mockRelease])
    releaseServiceMocks.getAllFileIds.mockResolvedValueOnce([])

    await expect(
      // @ts-expect-error protected
      exporter.checkReleaseFiles(),
    ).resolves.toBeUndefined()
    expect(fileServiceMocks.getFilesByIds).toHaveBeenCalled()
    expect(fileServiceMocks.getTotalFileSize).not.toHaveBeenCalled()
  })

  test('checkReleaseFiles > throws BadReq if total size exceeds max', async () => {
    fileServiceMocks.getTotalFileSize.mockResolvedValueOnce(2000)
    const exporter = new DocumentsExporter(mockUser, mockModel, [mockRelease, mockRelease])

    await expect(
      // @ts-expect-error protected method
      exporter.checkReleaseFiles(),
    ).rejects.toEqual(
      BadReq('Requested export is too large.', { size: 2000, maxSize: configMocks.default.modelMirror.export.maxSize }),
    )
  })

  test('checkReleaseFiles > throws BadReq if AV scan issues', async () => {
    scannersMocks.default.info.mockReturnValue(true)
    const badFile = { id: 'f', name: 'name', avScan: [] }
    fileServiceMocks.getFilesByIds.mockResolvedValueOnce([badFile as any])
    const exporter = new DocumentsExporter(mockUser, mockModel, [mockRelease])

    await expect(
      // @ts-expect-error protected method
      exporter.checkReleaseFiles(),
    ).rejects.toEqual(
      BadReq('The releases contain file(s) that do not have a clean AV scan.', {
        scanErrors: expect.any(Object),
      }),
    )
  })

  test('checkReleaseFiles > throws BadReq when AV scan incomplete', async () => {
    scannersMocks.default.info.mockReturnValue(true)
    const incompleteFile = { id: 'f', name: 'name', avScan: [{ state: ScanState.InProgress, isInfected: false }] }
    fileServiceMocks.getFilesByIds.mockResolvedValueOnce([incompleteFile as any])
    const exporter = new DocumentsExporter(mockUser, mockModel, [mockRelease])

    await expect(
      // @ts-expect-error protected
      exporter.checkReleaseFiles(),
    ).rejects.toEqual(
      BadReq('The releases contain file(s) that do not have a clean AV scan.', {
        scanErrors: expect.any(Object),
      }),
    )
  })

  test('checkReleaseFiles > throws BadReq when AV scan failed', async () => {
    scannersMocks.default.info.mockReturnValue(true)
    const infectedFile = { id: 'f', name: 'name', avScan: [{ state: ScanState.Complete, isInfected: true }] }
    fileServiceMocks.getFilesByIds.mockResolvedValueOnce([infectedFile as any])
    const exporter = new DocumentsExporter(mockUser, mockModel, [mockRelease])

    await expect(
      // @ts-expect-error protected
      exporter.checkReleaseFiles(),
    ).rejects.toEqual(
      BadReq('The releases contain file(s) that do not have a clean AV scan.', {
        scanErrors: expect.any(Object),
      }),
    )
  })

  test('getInitialiseTarGzUploadParams > returns correct params', () => {
    const exporter = new DocumentsExporter(mockUser, mockModel)

    // @ts-expect-error protected method
    const params = exporter.getInitialiseTarGzUploadParams()

    expect(params[0]).toBe(`${mockModel.id}.tar.gz`)
    expect(params[1]).toMatchObject({
      exporter: mockUser.dn,
      sourceModelId: mockModel.id,
      mirroredModelId: mockModel.settings.mirror.destinationModelId,
      importKind: MirrorKind.Documents,
    })
  })

  test('getInitialiseTarGzUploadParams > throws if model missing', () => {
    const exporter = new DocumentsExporter(mockUser, undefined as any)

    // @ts-expect-error protected method
    expect(() => exporter.getInitialiseTarGzUploadParams()).toThrowError(
      InternalError('Method `getInitialiseTarGzUploadParams` called before `this.model` defined.', {
        exporterType: 'DocumentsExporter',
      }),
    )
  })

  test('addData > throws if not initialised', () => {
    const exporter = new DocumentsExporter(mockUser, mockModel)

    expect(() => exporter.addData()).toThrowError(
      InternalError('Method `DocumentsExporter.addData` called before `init()`.', {
        exporterType: 'DocumentsExporter',
      }),
    )
  })

  test('addData > success calls addModelCardRevisionsToTarball and addReleasesToTarball', async () => {
    const exporter = new DocumentsExporter(mockUser, mockModel, [mockRelease])
    exporter['initialised'] = true
    exporter['tarStream'] = new PassThrough() as any
    exporter['gzipStream'] = new PassThrough() as any
    exporter['uploadStream'] = new PassThrough() as any
    exporter['uploadPromise'] = Promise.resolve()

    await exporter.addData()

    expect(modelServiceMocks.getModelCardRevisions).toHaveBeenCalled()
    expect(tarballMocks.addEntryToTarGzUpload).toHaveBeenCalled()
  })

  test('addData > wraps error from addModelCardRevisionsToTarball with InternalError message', async () => {
    tarballMocks.addEntryToTarGzUpload.mockImplementationOnce(() => {
      throw new Error('fail')
    })
    const exporter = new DocumentsExporter(mockUser, mockModel)
    exporter['initialised'] = true
    exporter['tarStream'] = new PassThrough() as any
    exporter['gzipStream'] = new PassThrough() as any
    exporter['uploadStream'] = new PassThrough() as any
    exporter['uploadPromise'] = Promise.resolve()

    await expect(exporter.addData()).rejects.toThrow(
      'Error when adding the model card revision(s) to the Tarball file.',
    )
  })

  test('addData > wraps error from addReleasesToTarball with InternalError message', async () => {
    const exporter = new DocumentsExporter(mockUser, mockModel, [mockRelease])
    exporter['initialised'] = true
    exporter['tarStream'] = new PassThrough() as any
    exporter['gzipStream'] = new PassThrough() as any
    exporter['uploadStream'] = new PassThrough() as any
    exporter['uploadPromise'] = Promise.resolve()
    const spy = vi.spyOn(exporter as any, 'addReleasesToTarball').mockImplementationOnce(() => {
      throw new Error('fail')
    })

    await expect(exporter.addData()).rejects.toThrow('Error when adding the release(s) to the Tarball file.')

    spy.mockRestore()
  })

  test('addModelCardRevisionsToTarball > propagates error from addEntryToTarGzUpload', async () => {
    tarballMocks.addEntryToTarGzUpload.mockImplementationOnce(() => {
      throw new Error('fail')
    })
    const exporter = new DocumentsExporter(mockUser, mockModel)
    exporter['initialised'] = true
    exporter['tarStream'] = new PassThrough() as any
    exporter['gzipStream'] = new PassThrough() as any
    exporter['uploadStream'] = new PassThrough() as any
    exporter['uploadPromise'] = Promise.resolve()

    await expect(
      // @ts-expect-error protected method
      exporter.addModelCardRevisionsToTarball(),
    ).rejects.toThrow('fail')
  })

  test('addReleasesToTarball > throws InternalError if release fails', async () => {
    const exporter = new DocumentsExporter(mockUser, mockModel, [mockRelease])
    exporter['initialised'] = true
    exporter['tarStream'] = new PassThrough() as any
    exporter['gzipStream'] = new PassThrough() as any
    exporter['uploadStream'] = new PassThrough() as any
    exporter['uploadPromise'] = Promise.resolve()
    const errExporter = vi.spyOn(exporter as any, 'addReleaseToTarball')
    errExporter.mockRejectedValueOnce(new Error('fail'))

    await expect(
      // @ts-expect-error protected method
      exporter.addReleasesToTarball(),
    ).rejects.toThrow('Error when adding release(s) to Tarball file.')
  })

  test('addReleaseToTarball > throws InternalError if entry fails', async () => {
    tarballMocks.addEntryToTarGzUpload.mockImplementationOnce(() => {
      throw new Error('fail')
    })
    const exporter = new DocumentsExporter(mockUser, mockModel, [mockRelease])
    exporter['initialised'] = true
    exporter['tarStream'] = new PassThrough() as any
    exporter['gzipStream'] = new PassThrough() as any
    exporter['uploadStream'] = new PassThrough() as any
    exporter['uploadPromise'] = Promise.resolve()

    await expect(
      // @ts-expect-error protected method
      exporter.addReleaseToTarball(mockRelease),
    ).rejects.toThrow('Error when generating the tarball file.')
  })

  test('addFilesToTarball > throws InternalError if entry fails', async () => {
    tarballMocks.addEntryToTarGzUpload.mockImplementationOnce(() => {
      throw new Error('fail')
    })
    const exporter = new DocumentsExporter(mockUser, mockModel)
    exporter['initialised'] = true
    exporter['tarStream'] = new PassThrough() as any
    exporter['gzipStream'] = new PassThrough() as any
    exporter['uploadStream'] = new PassThrough() as any
    exporter['uploadPromise'] = Promise.resolve()

    await expect(
      // @ts-expect-error protected method
      exporter.addFilesToTarball([mockFile]),
    ).rejects.toThrow('Error when generating the tarball file.')
  })
})
