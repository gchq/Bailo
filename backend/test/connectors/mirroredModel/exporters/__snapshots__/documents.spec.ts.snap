// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`connectors > mirroredModel > exporters > DocumentsExporter > constructor > sets releases and logData 1`] = `
DocumentsExporter {
  "_init": [Function],
  "addData": [Function],
  "addFilesToTarball": [Function],
  "addModelCardRevisionsToTarball": [Function],
  "addReleaseToTarball": [Function],
  "addReleasesToTarball": [Function],
  "checkReleaseFiles": [Function],
  "files": undefined,
  "getFiles": [Function],
  "getInitialiseTarGzUploadParams": [Function],
  "getReleases": [Function],
  "getSemvers": [Function],
  "gzipStream": undefined,
  "initialised": false,
  "logData": {
    "exportId": "exportId",
    "exporterType": "DocumentsExporter",
    "extra": "info",
  },
  "model": {
    "card": {
      "schemaId": "schemaId",
    },
    "id": "modelId",
    "settings": {
      "mirror": {
        "destinationModelId": "destModelId",
      },
    },
  },
  "releases": [
    {
      "foo": "bar",
    },
  ],
  "tarStream": undefined,
  "uploadPromise": undefined,
  "uploadStream": undefined,
  "user": {
    "dn": "userDN",
  },
}
`;
